### DISCUSSION TITLE
Why layout.tsx is executed when user can't access page?

### DISCUSSION THREAD

**Discussion by brunouber** [vercel/next.js#53526] [App Router]
Consider this file structure:

```
/app/path1/path2/
 - layout.tsx
 - page.tsx
```

and page being defined as:

```tsx
//path2/page.tsx

const Page = () => {
    return <div>...</div>;
};

export default withPageAuthRequired(Page)
```

`withPageAuthRequired` is method that check whether user can access this section, and otherwise it returns a `redirect()` to login page... you can imagine `withPageAuthRequired` being similar to the method in nextjs-auth0 library.

What I notice is that that `/app/path1/path2/layout.tsx` is executed even in case the user should not be allowed to access `path2/page.tsx`. What's the reason for that? Is there any way I could avoit this?

**Comment by Nadeera3784** [comment]
Hi [brunouber](https://github.com/brunouber),

How about wrapping the Layout component with withPageAuthRequired?

for example : 

```
// /app/path1/path2/layout.tsx

const Layout = () => {
    return <div>...</div>;
};

export default withPageAuthRequired(Layout);
```

```
// /app/path1/path2/page.tsx

const Page = () => {
    return <div>...</div>;
};

export default Page;


```

**Reply by brunouber** [reply]
Yes, I tried that, and it works - i mean, it prevents the layout from executing, but it has the problem that in the layout file I don't know the route url, so I can't implement proper returnTo behavior:

```
withPageAuthRequired(Layout, { returnTo: "???" };)
```

**Reply by Nadeera3784** [reply]
You can simply use

{ returnTo: 'http://localhost:3000/dashboard' }
{ returnTo: `${baseUrl}/${ctx.req.url}` }

**Reply by brunouber** [reply]
I see you're trying to help here, and I appreciate that - but I don't see how that can work.
I'm using app router, and afaik from a layout component there's now way to determine which exact path has been called (right?).
This is how my Layout component looks:

```
type AppPageProps = {
    params?: Record<string, string>;
    searchParams?: {
        [key: string]: string | string[] | undefined;
    };
};

const Layout = ({ children }) => {
    return <div className="fancy-layout">{children}</div>
}

export default onlyAuthenticated(Layout, { returnTo: (p: AppPageProps) => `??` });
```

Notice, how i updated a bit also the implementation of `onlyAuthenticated`, so that `returnTo` can either be the hardcoded destination url, or a function that returns it... in this latter case it can also receive `AppPageProps` as parameter - and this is the best I can do, but there's no way - again, i repeat myself (afaik) - to retrieve here the exact current path.

Am i missing something.
